buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java

```diff
@@ -0,0 +1,286 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.test;
+
+import org.elasticsearch.gradle.BuildPlugin;
+import org.elasticsearch.gradle.BwcVersions;
+import org.elasticsearch.gradle.DistributionDownloadPlugin;
+import org.elasticsearch.gradle.ElasticsearchDistribution;
+import org.elasticsearch.gradle.ElasticsearchDistribution.Flavor;
+import org.elasticsearch.gradle.ElasticsearchDistribution.Platform;
+import org.elasticsearch.gradle.ElasticsearchDistribution.Type;
+import org.elasticsearch.gradle.Jdk;
+import org.elasticsearch.gradle.JdkDownloadPlugin;
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.VersionProperties;
+import org.elasticsearch.gradle.info.GlobalBuildInfoPlugin;
+import org.elasticsearch.gradle.vagrant.BatsProgressLogger;
+import org.elasticsearch.gradle.vagrant.VagrantBasePlugin;
+import org.elasticsearch.gradle.vagrant.VagrantExtension;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.file.Directory;
+import org.gradle.api.plugins.ExtraPropertiesExtension;
+import org.gradle.api.plugins.JavaBasePlugin;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Copy;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.testing.Test;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.gradle.vagrant.VagrantService.convertLinuxPath;
+import static org.elasticsearch.gradle.vagrant.VagrantService.convertWindowsPath;
+
+public class DistroTestPlugin implements Plugin<Project> {
+
+    private static final String SYSTEM_JDK_VERSION = "11.0.2+9";
+    private static final String GRADLE_JDK_VERSION = "12.0.1+12@69cfe15208a647278a19ef0990eea691";
+
+    // all distributions used by distro tests. this is temporary until tests are per distribution
+    private static final String PACKAGING_DISTRIBUTION = "packaging";
+    private static final String COPY_PACKAGING_TASK = "copyPackagingArchives";
+    private static final String IN_VM_SYSPROP = "tests.inVM";
```

Comment:
This sysprop doesn't have anything to do with allowing destructive tasks to run. I added "destructive" to the task name so that it was clear this should not be run by a dev. Rather, this property just tells us whether the destructive task should have a dependency on the copyPackagingArchives task. When the sysprop is not present (ie when we will run from GCP for CI in a dedicated worker per platform) the dependency is added, thus causing the distributions to be built, rather than expecting them to already exist.

Long term I discussed a better plan with @mark-vieira to make the distribution plugin allow a sysprop to use a local file for a distribution. This will allow us to build the distribution outside the VM, then pass the path in via a sysprop, and gradle inside the VM will resolve the distro to the local file without trying to build it, yet the dependencies of the outer and inner tasks will all be the same (the configuration setup by the distribution plugin). But making that change requires first converting the tests to be distribution aware, ie passing in the distribution instead of expecting all of them to exist in a single directory.